### ISSUE TITLE
Expose version API in a consumable way

### ISSUE THREAD

**Issue by reuben** [mozilla/DeepSpeech#2687]
Currently we have `DS_PrintVersions` which is basically just a convenience function for our clients. As we're working towards API stabilization and things like a model downloader/installer, it becomes useful to be able to utilize the version information from the API.

Simplest possible solution is to just return the version string instead of printing it. Then API consumers would have to parse the result. Another possibility is to do something like what the TensorFlow Python module has, and separate the information like this (eg. for DeepSpeech v0.6.1):

VERSION = '0.6.1'
TF_VERSION = '1.14.0'
GIT_VERSION = 'v0.6.1-0-g3df20fe'
TF_GIT_VERSION = 'v1.14.0-21-ge77504ac6b'

This means the semver-style info is easily available in the _VERSION constants, and the full git version can be used by clients/diagnostics if needed.

One example use case: we write a downloader tool for models, and it can check the version of the model being downloaded against the version of the `deepspeech` package that's installed, warning you if things are not compatible.

**Comment by reuben** [comment]
@lissyx thoughts?

**Comment by lissyx** [comment]
> @lissyx thoughts?

Having work with / on packaging at distro level, I know parsing versions can be a source of issues.
I think the proposal make sense, but I would like to avoid exposing as strings, at least for our version.

We could have `VERSION=0x000601` for 0.6.1, this makes it easy to read as well as to use for comparisions.

I don't think we should expose TensorFlow version, though.

Someone reported that exposing the underlying runtime might be of help as well, maybe we should to that at the same time ?

Like:
```
enum {
    TENSORFLOW_CPU = 0x00,
    TENSORFLOW_GPU_CUDA = 0x01,
    TENSORFLOW_TFLITE = 0x02,
} _runtime;
```

?

**Comment by reuben** [comment]
> Having work with / on packaging at distro level, I know parsing versions can be a source of issues.

Even with semver? There's probably a library to parse semver version strings for every language out there.

> We could have VERSION=0x000601 for 0.6.1, this makes it easy to read as well as to use for comparisions.

I'd rather avoid having to ship version `B.A.D`. How would you encode `v0.7.0-alpha.0` in that scheme? I also don't see how this makes it easier to compare? One would still need to extract the major, minor, patch, pre-release components out of the number and then apply the same logic.

If parsing is a problem and we need to expose individual numerical components I'd rather go with

```
VERSION_MAJOR = 0
VERSION_MINOR = 6
VERSION_PATCH = 1
VERSION_PRERELEASE = ''
```

> Someone reported that exposing the underlying runtime might be of help as well, maybe we should to that at the same time ?

Agreed, that would be a good thing to have.

**Comment by lissyx** [comment]
> I'd rather avoid having to ship version `B.A.D`. How would you encode `v0.7.0-alpha.0` in that scheme? I also don't see how this makes it easier to compare? One would still need to extract the major, minor, patch, pre-release components out of the number and then apply the same logic.

Do we care about the alpha part? In this case, `VERSION=0x000700` and that's it



> I also don't see how this makes it easier to compare?

Correct me if I'm wrong, but `0x000601 < 0x000700`, while it's a bit more painful when you separate MAJOR/MINOR/PATCH.

**Comment by lissyx** [comment]
> ```
> 
> ```
> 
> 
> > Someone reported that exposing the underlying runtime might be of help as well, maybe we should to that at the same time ?
> 
> 
> Agreed, that would be a good thing to have.

He was referring to some automation regarding the filenames for the model. Maybe we could avoid exposing that information and just be more intelligent ?

I'm afraid of exposing too much and we have to deal with strange compatibility workaround in the future ?

**Comment by lissyx** [comment]
> > Having work with / on packaging at distro level, I know parsing versions can be a source of issues.
> 
> Even with semver? There's probably a library to parse semver version strings for every language out there.

It's not wrong that semver do comes with much higher standards regarding this problem. Maybe we could live with just that, though I find it a bit painful if, e.g., I need a lib parse and handle `#ifdef` in plain `C`.

**Comment by reuben** [comment]
> Do we care about the alpha part? In this case, `VERSION=0x000700` and that's it

We currently ship our alpha builds, so code relying on checks like that wouldn't be able to differentiate between `v0.7.0-alpha.N` and `v0.7.0` stable. For our diagnostics we'd lose precious information (this person is using an alpha build, and which one).

**Comment by reuben** [comment]
I do see your point that in C code with a single integer you could then easily do something like:

```c
int version = DS_Version();
int minimum_supported = 0x000600;
int scorer_package = 0x000700;
int runs_on_my_fridge = 0x000800;
if (version <= minimum_supported) {
    DS_EnableDecoderWithLM(...);
} else if (version <= scorer_package) {
    DS_EnableExternalScorer(...);
} else if (version <= runs_on_my_fridge) {
    DS_EnableGroceriesDetection(...);
} // ...and so on
```

**Comment by reuben** [comment]
I think because we are using semver, we should use it consistently, and expose the suffix as well.

**Comment by reuben** [comment]
> He was referring to some automation regarding the filenames for the model. Maybe we could avoid exposing that information and just be more intelligent ?
> 
> I'm afraid of exposing too much and we have to deal with strange compatibility workaround in the future ?

Yeah, thinking more about it I feel like it's exposing too much of an implementation detail. Instead I think users should simply try to load models in order of preference and check for errors when loading. I think currently TF and TFLite implementations return different error codes for "format not understood", so we could make that consistent to help. For CUDA, I can't think of a reason why you'd want to programaticaly switch the input model depending on that...

**Comment by lissyx** [comment]
> I think currently TF and TFLite implementations return different error codes for "format not understood", so we could make that consistent to help.

We could, but I would also understand that people find this way to deal kind of not completely satifsying.



> For CUDA, I can't think of a reason why you'd want to programaticaly switch the input model depending on that...

For CUDA, I don't remember the rationale but there was something else, some time ago.

**Comment by lissyx** [comment]
@reuben So, SemVer string + C-level define ? We should be fine ?

**Comment by reuben** [comment]
@lissyx I don't like doing both at the same time, and I'm not sure how useful C `#define`s are since you can't easily handle the pre-release part. I'm tempted to simply expose the version strings (DS and TF) at runtime: `char* DS_GetVersion();` and `char* DS_GetTFVersion();`.

**Comment by lissyx** [comment]
> I'm not sure how useful C `#define`s are since you can't easily handle the pre-release part

I'm really not sure it's a big deal.



> I'm tempted to simply expose the version strings (DS and TF) at runtime: `char* DS_GetVersion();` and `char* DS_GetTFVersion();`.

Let's go with that, but I'm not sure we should expose TensorFlow version: our library makes it completely hidden, and our deepspeech versioning scheme already handles incompatibilities that might arise from TensorFlow.

What makes you think we should expose it?

**Comment by reuben** [comment]
> I'm really not sure it's a big deal.

Whatever we expose in 1.0 we have to live with during 1.x. If we expose two different versions, one SemVer and one truncated, we'll have to maintain both. I'd rather stick to the versioning scheme we've been using so far, which is SemVer.

> What makes you think we should expose it?

Same reason it's currently printed by `DS_PrintVersions`, diagnostics. If you're OK with dropping it I'm OK with it too.

**Comment by lissyx** [comment]
> > I'm really not sure it's a big deal.
> 
> Whatever we expose in 1.0 we have to live with during 1.x. If we expose two different versions, one SemVer and one truncated, we'll have to maintain both. I'd rather stick to the versioning scheme we've been using so far, which is SemVer.

Agreed, it's just that it's easy to do. Let's keep SemVer only, we'll see if people need more.



> > What makes you think we should expose it?
> 
> Same reason it's currently printed by `DS_PrintVersions`, diagnostics. If you're OK with dropping it I'm OK with it too.

Right, but I'm unsure if it makes sense to do that API-level, since people **might** not share it ahead of time and we get screwed. And having to keep that exposed means we have to keep it for a while :D.

I don't have a very strong opinion here :/

**Comment by kdavis-mozilla** [comment]
Should we put this in 0.7.0 or 1.0.0? Or somewhere else? (I'm leaning towards 0.7.0)

**Comment by lissyx** [comment]
I vote for 0.7.0 so we can get feedback from the ground

**Comment by reuben** [comment]
Yes, specially as it's going to take a some time until the new training run with the MFCC fix, which gives us time to fix this without holding the release with a model ready.

**Comment by reuben** [comment]
Fixed by #2745
